File: cpp/src/Perlin.cpp
#include "Perlin.hpp"
#include "RandomVector.hpp"

Perlin::Perlin(int seed) : m_vector(seed) {
    m_seed = seed;
    std::cout << "Initialising Perlin Generator with seed: " << seed << std::endl;
}

float Perlin::operator()(float xp, float yp){
    int xlower = int(xp);
    int ylower = int(yp);
    int xupper = xlower + 1;
    int yupper = ylower + 1;

    float dx = Perlin::fade(xp - float(xlower));
    float dy = Perlin::fade(yp - float(ylower));

    float rgLower0 = m_vector.eval(xp, yp, xlower, ylower);
    float rgUpper0 = m_vector.eval(xp, yp, xupper, ylower);
    float interpLower = Perlin::lerp(rgLower0, rgUpper0, dx);

    float rgLower1 = m_vector.eval(xp, yp, xlower, yupper);
    float rgUpper1 = m_vector.eval(xp, yp, xupper, yupper);
    float interpUpper = Perlin::lerp(rgLower1, rgUpper1, dx);

    return Perlin::lerp(interpLower, interpUpper, dy);
}

std::vector<float> Perlin::rescale(std::vector<float> input) {
    float _max = *std::max_element(input.begin(), input.end());
    float _min = *std::min_element(input.begin(), input.end());

    std::vector<float> _output;
    _output.reserve(input.size());

    for (int i = 0; i < input.size(); i++) {
        _output.push_back(-1 + (input.at(i) - _min) / (_max - _min) * 2);
    }
    return _output;
}

File: perlin_rust/src/perlin.rs
#[path = "random_vector.rs"] mod random_vector;


pub struct Perlin {
    rnd_vec: random_vector::RandomVector,
}

fn fade(val : f32) -> f32 {
    val * val * val * ( val * ( val * 6.0 - 15.0 ) + 10.0 )
}

fn lerp(v0: f32, v1 : f32, ds : f32) -> f32 {
    v0 + (v1 - v0) * ds
}

impl Perlin {
    pub fn new(seed: u64) -> Perlin {
        println!("Initialising Perlin creator with seed: {}", seed);
        let rnd_vec = random_vector::RandomVector::new(seed);
        Perlin{ rnd_vec }
    }

    pub fn eval(&mut self, xp: f32, yp: f32) -> f32 {
        let xlower : i32 = xp as i32;
        let ylower : i32 = yp as i32;
        let xupper : i32 = xlower + 1;
        let yupper : i32 = ylower + 1;

        let dx : f32 = fade(xp - xlower as f32);
        let dy : f32 = fade(yp - ylower as f32);

        let rg_lower_0 = self.rnd_vec.eval(xp, yp, xlower, ylower);
        let rg_upper_0 = self.rnd_vec.eval(xp, yp, xupper, ylower);
        let interp_lower = lerp(rg_lower_0, rg_upper_0, dx);

        let rg_lower_1 = self.rnd_vec.eval(xp, yp, xlower, yupper);
        let rg_upper_1 = self.rnd_vec.eval(xp, yp, xupper, yupper);
        let interp_upper = lerp(rg_lower_1, rg_upper_1, dx);

        lerp(interp_lower, interp_upper, dy)
    }

}